rclcpp/include/rclcpp/strategies/msg_pool_memory_strategy.hpp

```diff
@@ -0,0 +1,80 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP_RCLCPP_MSG_POOL_MEMORY_STRATEGY_HPP_
+#define RCLCPP_RCLCPP_MSG_POOL_MEMORY_STRATEGY_HPP_
+
+#include <rclcpp/msg_memory_strategy.hpp>
+#include <rclcpp/macros.hpp>
+
+namespace rclcpp
+{
+namespace msg_pool_memory_strategy
+{
+
+template<typename MessageT, size_t size>
+class MessagePoolMemoryStrategy
+  : public std::enable_if<rosidl_generator_traits::has_fixed_size<MessageT>::value,
+  msg_memory_strategy::MessageMemoryStrategy<MessageT>>::type
+{
+public:
+  RCLCPP_MAKE_SHARED_DEFINITIONS(MessagePoolMemoryStrategy);
+  MessagePoolMemoryStrategy()
+  : seq_(0)
+  {
+    for (size_t i = 0; i < size; ++i) {
+      msg_ptrs_[i] = std::make_shared<MessageT>(msg_pool_[i]);
+    }
+
+    msg_used_.fill(false);
+  }
+
+  std::shared_ptr<MessageT> borrow_message()
+  {
+    seq_ = (seq_ + 1) % size;
```

Comment:
It is not intuitive that with the current logic the ring buffer return the first borrowed message from index 1 instead of 0.